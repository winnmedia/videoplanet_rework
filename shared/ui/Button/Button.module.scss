@use '@/shared/config/design-tokens.scss' as *;

.button {
  @include button-base;
  position: relative;
  overflow: hidden;
  font-family: $font-primary;
  
  // Ripple effect
  &::before {
    content: '';
    position: absolute;
    top: 50%;
    left: 50%;
    width: 0;
    height: 0;
    border-radius: $radius-full;
    background: rgba(255, 255, 255, 0.5);
    transform: translate(-50%, -50%);
    transition: width 0.6s, height 0.6s;
  }
  
  &:active::before {
    width: 300px;
    height: 300px;
  }
}

// Sizes
.sm {
  height: $button-height-sm;
  padding: 0 $spacing-md;
  font-size: $font-size-sm;
  border-radius: $radius-xl;
}

.md {
  height: $button-height-md;
  padding: 0 $spacing-lg;
  font-size: $font-size-base;
  border-radius: $radius-xl;
}

.lg {
  height: $button-height-lg;
  padding: 0 $spacing-2xl;
  font-size: $font-size-md;
  border-radius: $radius-2xl;
}

// Variants
.primary {
  background: linear-gradient(135deg, $color-primary 0%, $color-primary-dark 100%);
  color: $color-white;
  
  @include hover-lift;
  
  &:hover:not(:disabled) {
    background: linear-gradient(135deg, $color-primary-dark 0%, $color-primary-darker 100%);
    box-shadow: $shadow-primary;
  }
  
  &:active:not(:disabled) {
    transform: translateY(0);
  }
}

.secondary {
  background: $color-white;
  color: $color-dark;
  border: $border-1 solid $color-gray-300;
  
  &:hover:not(:disabled) {
    background: $color-gray-100;
    border-color: $color-primary;
    color: $color-primary;
    transform: translateY(-1px);
    box-shadow: $shadow-sm;
  }
  
  &:active:not(:disabled) {
    transform: translateY(0);
  }
}

.outline {
  background: transparent;
  color: $color-primary;
  border: $border-2 solid $color-primary;
  
  &:hover:not(:disabled) {
    background: rgba($color-primary, 0.05);
    transform: translateY(-1px);
    box-shadow: $shadow-sm;
  }
  
  &:active:not(:disabled) {
    transform: translateY(0);
  }
  
  &::before {
    background: rgba($color-primary, 0.2);
  }
}

.danger {
  background: linear-gradient(135deg, $color-error 0%, $color-error-dark 100%);
  color: $color-white;
  
  @include hover-lift;
  
  &:hover:not(:disabled) {
    background: linear-gradient(135deg, $color-error-dark 0%, $color-error-darker 100%);
    box-shadow: 0 5px 20px rgba($color-error, 0.3);
  }
  
  &:active:not(:disabled) {
    transform: translateY(0);
  }
}

.ghost {
  background: transparent;
  color: $color-dark;
  
  &:hover:not(:disabled) {
    background: rgba($color-dark, 0.05);
  }
  
  &:active:not(:disabled) {
    background: rgba($color-dark, 0.1);
  }
}

// States
.fullWidth {
  width: 100%;
}

.loading {
  color: transparent;
  pointer-events: none;
  
  &::after {
    content: '';
    position: absolute;
    width: 20px;
    height: 20px;
    top: 50%;
    left: 50%;
    margin-left: -10px;
    margin-top: -10px;
    border: 2px solid $color-white;
    border-radius: 50%;
    border-top-color: transparent;
    animation: spinner 0.8s linear infinite;
  }
  
  &.secondary::after,
  &.outline::after,
  &.ghost::after {
    border-color: $color-primary;
    border-top-color: transparent;
  }
  
  @keyframes spinner {
    to { transform: rotate(360deg); }
  }
}

// Icon support
.icon {
  display: inline-flex;
  align-items: center;
  gap: $spacing-sm;
  
  svg {
    width: 1.2em;
    height: 1.2em;
  }
}

// Responsive
@include mobile {
  .sm {
    height: 32px;
    padding: 0 $spacing-sm;
    font-size: $font-size-xs;
  }
  
  .md {
    height: 40px;
    padding: 0 $spacing-md;
    font-size: $font-size-sm;
  }
  
  .lg {
    height: 48px;
    padding: 0 $spacing-lg;
    font-size: $font-size-base;
  }
}