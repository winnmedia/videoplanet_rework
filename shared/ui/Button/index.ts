/**
 * @fileoverview Button 컴포넌트 Public API - 레거시 지원
 * @description 기존 코드와의 하위 호환성을 위한 레거시 Button Export
 */

export { Button } from './Button.modern'
export type { ButtonProps } from './Button.modern'
