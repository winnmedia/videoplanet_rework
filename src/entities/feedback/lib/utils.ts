import { FeedbackStatus, FeedbackPriority, FeedbackType } from '../model/types'

/**
 * 상태별 표시 설정
 */
export const FEEDBACK_STATUS_CONFIG = {
  pending: {
    label: '대기중',
    color: 'bg-admin-warning text-white',
    description: '검토 대기중인 피드백',
  },
  approved: {
    label: '승인됨',
    color: 'bg-admin-success text-white',
    description: '승인된 피드백',
  },
  rejected: {
    label: '거부됨',
    color: 'bg-admin-error text-white',
    description: '거부된 피드백',
  },
  implemented: {
    label: '구현완료',
    color: 'bg-primary-600 text-white',
    description: '구현이 완료된 피드백',
  },
} as const

/**
 * 우선순위별 표시 설정
 */
export const FEEDBACK_PRIORITY_CONFIG = {
  low: {
    label: '낮음',
    color: 'bg-neutral-200 text-neutral-700',
    sortOrder: 1,
  },
  medium: {
    label: '보통',
    color: 'bg-blue-100 text-blue-700',
    sortOrder: 2,
  },
  high: {
    label: '높음',
    color: 'bg-amber-100 text-amber-700',
    sortOrder: 3,
  },
  urgent: {
    label: '긴급',
    color: 'bg-red-100 text-red-700',
    sortOrder: 4,
  },
} as const

/**
 * 유형별 표시 설정
 */
export const FEEDBACK_TYPE_CONFIG = {
  general: {
    label: '일반',
    color: 'bg-neutral-100 text-neutral-700',
    icon: '💬',
  },
  bug: {
    label: '버그',
    color: 'bg-red-100 text-red-700',
    icon: '🐛',
  },
  feature: {
    label: '기능요청',
    color: 'bg-green-100 text-green-700',
    icon: '✨',
  },
  improvement: {
    label: '개선사항',
    color: 'bg-blue-100 text-blue-700',
    icon: '📈',
  },
  question: {
    label: '질문',
    color: 'bg-purple-100 text-purple-700',
    icon: '❓',
  },
} as const

/**
 * 상태의 표시 정보를 가져옵니다
 */
export function getStatusConfig(status: FeedbackStatus) {
  return FEEDBACK_STATUS_CONFIG[status]
}

/**
 * 우선순위의 표시 정보를 가져옵니다
 */
export function getPriorityConfig(priority: FeedbackPriority) {
  return FEEDBACK_PRIORITY_CONFIG[priority]
}

/**
 * 유형의 표시 정보를 가져옵니다
 */
export function getTypeConfig(type: FeedbackType) {
  return FEEDBACK_TYPE_CONFIG[type]
}

/**
 * 우선순위를 기준으로 정렬 순서를 반환합니다
 */
export function getPrioritySortOrder(priority: FeedbackPriority): number {
  return FEEDBACK_PRIORITY_CONFIG[priority].sortOrder
}

/**
 * 두 우선순위를 비교합니다 (높은 우선순위가 먼저)
 */
export function comparePriorities(a: FeedbackPriority, b: FeedbackPriority): number {
  return getPrioritySortOrder(b) - getPrioritySortOrder(a)
}

/**
 * 피드백이 긴급한지 확인합니다
 */
export function isUrgentFeedback(priority: FeedbackPriority): boolean {
  return priority === 'urgent' || priority === 'high'
}

/**
 * 피드백이 완료되었는지 확인합니다
 */
export function isCompletedFeedback(status: FeedbackStatus): boolean {
  return status === 'implemented' || status === 'rejected'
}

/**
 * 피드백이 처리 대기중인지 확인합니다
 */
export function isPendingFeedback(status: FeedbackStatus): boolean {
  return status === 'pending'
}

/**
 * 피드백 해결 시간을 계산합니다 (일 단위)
 */
export function calculateResolutionTime(created: string, resolvedAt?: string): number | null {
  if (!resolvedAt) return null
  
  const createdDate = new Date(created)
  const resolvedDate = new Date(resolvedAt)
  const diffTime = resolvedDate.getTime() - createdDate.getTime()
  const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24))
  
  return diffDays
}

/**
 * 피드백이 지연되었는지 확인합니다 (우선순위 기반)
 */
export function isOverdueFeedback(
  created: string,
  priority: FeedbackPriority,
  status: FeedbackStatus
): boolean {
  if (isCompletedFeedback(status)) return false
  
  const createdDate = new Date(created)
  const now = new Date()
  const daysSinceCreated = (now.getTime() - createdDate.getTime()) / (1000 * 60 * 60 * 24)
  
  // 우선순위별 SLA (일)
  const slaThresholds = {
    urgent: 1,
    high: 3,
    medium: 7,
    low: 14,
  }
  
  return daysSinceCreated > slaThresholds[priority]
}

/**
 * 타임스탬프를 사람이 읽기 쉬운 형식으로 변환합니다
 */
export function formatTimestamp(seconds?: number): string {
  if (!seconds) return '-'
  
  const hours = Math.floor(seconds / 3600)
  const minutes = Math.floor((seconds % 3600) / 60)
  const remainingSeconds = Math.floor(seconds % 60)
  
  if (hours > 0) {
    return `${hours}:${minutes.toString().padStart(2, '0')}:${remainingSeconds.toString().padStart(2, '0')}`
  }
  
  return `${minutes}:${remainingSeconds.toString().padStart(2, '0')}`
}

/**
 * 파일 크기를 사람이 읽기 쉬운 형식으로 변환합니다
 */
export function formatFileSize(bytes: number): string {
  const units = ['B', 'KB', 'MB', 'GB']
  let size = bytes
  let unitIndex = 0
  
  while (size >= 1024 && unitIndex < units.length - 1) {
    size /= 1024
    unitIndex++
  }
  
  return `${size.toFixed(unitIndex === 0 ? 0 : 1)}${units[unitIndex]}`
}

/**
 * 피드백 제목을 요약합니다
 */
export function truncateFeedbackTitle(title: string, maxLength: number = 50): string {
  if (title.length <= maxLength) return title
  return `${title.slice(0, maxLength)}...`
}

/**
 * 태그를 정규화합니다 (소문자, 공백 제거)
 */
export function normalizeTag(tag: string): string {
  return tag.toLowerCase().trim().replace(/\s+/g, '-')
}

/**
 * 태그 목록을 정규화하고 중복을 제거합니다
 */
export function normalizeTags(tags: string[]): string[] {
  const normalized = tags.map(normalizeTag).filter(tag => tag.length > 0)
  return Array.from(new Set(normalized))
}