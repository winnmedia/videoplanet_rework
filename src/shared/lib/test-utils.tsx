/**
 * 테스트 유틸리티
 * RTK Query와 Redux를 위한 테스트 환경 설정
 */

import React, { PropsWithChildren } from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import { server } from '../api/mocks/server'
import { planningApi } from '../../features/video-planning/api/planningApi'

// Test store configuration
export const createTestStore = (preloadedState?: any) => {
  return configureStore({
    reducer: {
      // RTK Query API slice
      [planningApi.reducerPath]: planningApi.reducer,
    },
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        // RTK Query에서 직렬화 검사를 비활성화 (테스트에서는 불필요)
        serializableCheck: {
          ignoredActions: [
            'persist/PERSIST',
            'persist/REHYDRATE',
            'persist/PAUSE',
            'persist/PURGE',
            'persist/REGISTER',
          ],
        },
      }).concat(planningApi.middleware),
  })
}

// Test wrapper with RTK Query provider
interface TestWrapperProps extends PropsWithChildren {
  store?: ReturnType<typeof createTestStore>
}

const TestWrapper: React.FC<TestWrapperProps> = ({ 
  children, 
  store = createTestStore() 
}) => {
  return (
    <Provider store={store}>
      {children}
    </Provider>
  )
}

// Custom render function with RTK Query support
export const renderWithProviders = (
  ui: React.ReactElement,
  {
    store = createTestStore(),
    ...renderOptions
  }: {
    store?: ReturnType<typeof createTestStore>
  } & Omit<Parameters<typeof rtlRender>[1], 'wrapper'> = {}
) => {
  const Wrapper: React.FC<PropsWithChildren> = ({ children }) => (
    <TestWrapper store={store}>{children}</TestWrapper>
  )

  return {
    store,
    ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }),
  }
}

// MSW server setup for tests
export const setupMSW = () => {
  // MSW 서버 설정
  beforeAll(() => {
    server.listen({ onUnhandledRequest: 'error' })
  })

  afterEach(() => {
    server.resetHandlers()
  })

  afterAll(() => {
    server.close()
  })
}

// RTK Query cache cleanup
export const cleanupRTKQuery = () => {
  afterEach(() => {
    // RTK Query 캐시 초기화
    planningApi.util.resetApiState()
  })
}

// Re-export everything from RTL
export * from '@testing-library/react'
export { renderWithProviders as render }