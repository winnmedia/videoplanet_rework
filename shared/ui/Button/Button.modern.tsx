/**
 * @fileoverview 초미니멀 Button 컴포넌트 - Tailwind CSS 기반
 * @description VRidge 디자인 시스템의 핵심 Button 컴포넌트
 */

'use client'

import { cva, type VariantProps } from 'class-variance-authority'
import { ButtonHTMLAttributes, ReactNode, forwardRef } from 'react'

import { cn } from '../../lib/utils'

// Button variants 정의 (CVA 사용)
const buttonVariants = cva(
  // VRidge 레거시 Button 완전 재현 (초미니멀 + 세련됨)
  [
    // 기본 구조 (@include button-base 매핑)
    'relative overflow-hidden',
    'inline-flex items-center justify-center',
    'font-semibold leading-none',
    'rounded-full border-none outline-none cursor-pointer',
    
    // VRidge 브랜드 전환 효과 (레거시 정확 매핑)
    'transition-all duration-200 ease-in-out',
    
    // 포커스 스타일 (레거시 @include focus-visible 매핑)
    'focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-vridge-500 focus-visible:ring-offset-2',
    
    // 액티브 상태 (레거시 scale 효과 정확 매핑)
    'active:scale-95 active:transition-transform active:duration-75',
    
    // 비활성화 상태
    'disabled:opacity-50 disabled:cursor-not-allowed disabled:active:scale-100',
    
    // 리플 효과를 위한 기본 설정 (레거시 &::before 매핑)
    'before:absolute before:inset-1/2 before:w-0 before:h-0 before:rounded-full',
    'before:bg-white/50 before:-translate-x-1/2 before:-translate-y-1/2',
    'before:transition-all before:duration-[600ms]',
    'active:before:w-[300px] active:before:h-[300px]'
  ],
  {
    variants: {
      variant: {
        primary: [
          // 레거시 그라데이션 완전 재현 (135deg, primary → primary-dark)
          'bg-gradient-to-br from-vridge-500 to-vridge-700',
          'text-white',
          
          // 호버 효과 (레거시 hover-lift + 그라데이션 변경)
          'hover:from-vridge-700 hover:to-vridge-800',
          'hover:shadow-primary hover:-translate-y-0.5',
          'transition-transform transition-shadow',
          
          // 액티브 상태 (translate 원복)
          'active:translate-y-0',
          
          // 포커스 링 색상
          'focus-visible:ring-vridge-500'
        ],
        secondary: [
          // 레거시 secondary 완전 재현
          'bg-white text-neutral-950',
          'border border-neutral-300',
          
          // 호버 효과 (레거시 정확 매핑)
          'hover:bg-neutral-100 hover:border-vridge-500 hover:text-vridge-500',
          'hover:-translate-y-px hover:shadow-sm',
          
          // 액티브 상태
          'active:translate-y-0',
          
          // 리플 효과 색상
          'before:bg-vridge-500/20',
          
          // 포커스 링
          'focus-visible:ring-neutral-500'
        ],
        outline: [
          // 레거시 outline 완전 재현
          'bg-transparent text-vridge-500',
          'border-2 border-vridge-500',
          
          // 호버 효과
          'hover:bg-vridge-500/5 hover:-translate-y-px hover:shadow-sm',
          
          // 액티브 상태
          'active:translate-y-0',
          
          // 리플 효과 (레거시 rgba($color-primary, 0.2))
          'before:bg-vridge-500/20',
          
          // 포커스 링
          'focus-visible:ring-vridge-500'
        ],
        danger: [
          // 레거시 danger 완전 재현
          'bg-gradient-to-br from-error-500 to-error-600',
          'text-white',
          
          // 호버 효과 (error shadow 매핑)
          'hover:from-error-600 hover:to-error-700',
          'hover:shadow-[0_5px_20px_rgba(217,58,58,0.3)] hover:-translate-y-0.5',
          
          // 액티브 상태
          'active:translate-y-0',
          
          // 포커스 링
          'focus-visible:ring-error-500'
        ],
        ghost: [
          // 레거시 ghost 완전 재현
          'bg-transparent text-neutral-950',
          
          // 호버 효과 (rgba 정확 매핑)
          'hover:bg-neutral-950/5',
          
          // 액티브 상태
          'active:bg-neutral-950/10',
          
          // 포커스 링
          'focus-visible:ring-neutral-500'
        ]
      },
      size: {
        sm: [
          // 레거시 .sm 정확 매핑
          'h-9 px-4 text-sm',        // $button-height-sm: 36px, $spacing-md: 16px, $font-size-sm: 14px
          'rounded-xl',              // $radius-xl: 16px (레거시)
          'gap-1.5',                 // 6px gap
          
          // 모바일 대응 (레거시 @include mobile 매핑)
          'sm:h-8 sm:px-2 sm:text-xs' // 32px, 8px, 12px (모바일)
        ],
        default: [
          // 레거시 .md 정확 매핑 (기본값)
          'h-11 px-6 text-base',     // $button-height-md: 44px, $spacing-lg: 24px, $font-size-base: 16px
          'rounded-xl',              // $radius-xl: 16px
          'gap-2',                   // 8px gap
          
          // 모바일 대응
          'sm:h-10 sm:px-4 sm:text-sm' // 40px, 16px, 14px (모바일)
        ],
        lg: [
          // 레거시 .lg 정확 매핑
          'h-14 px-10 text-lg',      // $button-height-lg: 54px → 56px 근사, $spacing-2xl: 40px, $font-size-md: 18px
          'rounded-2xl',             // $radius-2xl: 20px (레거시)
          'gap-2.5',                 // 10px gap
          
          // 모바일 대응
          'sm:h-12 sm:px-6 sm:text-base' // 48px, 24px, 16px (모바일)
        ]
      },
      fullWidth: {
        true: 'w-full'              // 레거시 .fullWidth 매핑
      }
    },
    defaultVariants: {
      variant: 'primary',
      size: 'default'
    }
  }
)

export interface ButtonProps 
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants> {
  /**
   * 버튼 내부에 표시될 콘텐츠
   */
  children: ReactNode
  
  /**
   * 로딩 상태 표시 여부
   */
  loading?: boolean
  
  /**
   * 아이콘 컴포넌트
   */
  icon?: ReactNode
  
  /**
   * 아이콘 위치
   */
  iconPosition?: 'left' | 'right'
  
  /**
   * 전체 너비 사용 여부
   */
  fullWidth?: boolean
}

/**
 * 로딩 스피너 컴포넌트
 */
const LoadingSpinner = () => (
  <svg
    role="status" 
    aria-hidden="true"
    className="w-4 h-4 animate-spin"
    viewBox="0 0 24 24"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      className="opacity-25"
      cx="12"
      cy="12"
      r="10"
      stroke="currentColor"
      strokeWidth="4"
    />
    <path
      className="opacity-75"
      fill="currentColor"
      d="M4 12a8 8 0 0 1 8-8v8H4z"
    />
  </svg>
)

/**
 * Button 컴포넌트
 * 
 * @description VRidge 디자인 시스템의 핵심 Button 컴포넌트
 * 초미니멀한 디자인과 완벽한 접근성을 제공합니다.
 * 
 * @example
 * ```tsx
 * <Button variant="primary" size="default">
 *   클릭하세요
 * </Button>
 * 
 * <Button variant="outline" loading>
 *   로딩 중...
 * </Button>
 * 
 * <Button variant="ghost" icon={<Icon />} iconPosition="right">
 *   아이콘과 함께
 * </Button>
 * ```
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ 
    className,
    variant,
    size,
    fullWidth,
    loading = false,
    icon,
    iconPosition = 'left',
    disabled,
    children,
    onClick,
    'aria-label': ariaLabel,
    ...props 
  }, ref) => {
    const isDisabled = disabled || loading
    
    const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
      if (isDisabled) {
        e.preventDefault()
        return
      }
      onClick?.(e)
    }

    const handleKeyDown = (e: React.KeyboardEvent<HTMLButtonElement>) => {
      if (isDisabled) return
      
      if (e.key === 'Enter' || e.key === ' ') {
        e.preventDefault()
        handleClick(e as any)
      }
    }

    return (
      <button
        ref={ref}
        className={cn(
          buttonVariants({ variant, size, fullWidth }),
          className
        )}
        disabled={isDisabled}
        aria-disabled={isDisabled}
        aria-busy={loading}
        aria-label={ariaLabel || (loading ? '처리 중' : undefined)}
        onClick={handleClick}
        onKeyDown={handleKeyDown}
        {...props}
      >
        {loading ? (
          <LoadingSpinner />
        ) : (
          <>
            {icon && iconPosition === 'left' && icon}
            <span className="truncate">{children}</span>
            {icon && iconPosition === 'right' && icon}
          </>
        )}
      </button>
    )
  }
)

Button.displayName = 'Button'