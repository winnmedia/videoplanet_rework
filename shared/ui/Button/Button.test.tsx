/**
 * Button Component Tests
 * Testing a shared UI component following TDD principles
 */

import { screen, fireEvent, waitFor } from '@testing-library/react'
import { describe, it, expect, vi } from 'vitest'

import { render, setup } from '@/test/utils/test-utils'

import { Button } from './Button'

describe('Button Component', () => {
  describe('Rendering', () => {
    it('should render with children text', () => {
      render(<Button>Click me</Button>)
      
      const button = screen.getByRole('button', { name: /click me/i })
      expect(button).toBeInTheDocument()
    })
    
    it('should render with default primary variant', () => {
      render(<Button>Primary Button</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'primary', 'md')
    })
    
    it('should render with secondary variant', () => {
      render(<Button variant="secondary">Secondary Button</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'secondary', 'md')
    })
    
    it('should render with danger variant', () => {
      render(<Button variant="danger">Delete</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'danger', 'md')
    })
  })
  
  describe('Sizes', () => {
    it('should render with medium size by default', () => {
      render(<Button>Medium Button</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'primary', 'md')
    })
    
    it('should render with small size', () => {
      render(<Button size="sm">Small Button</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'primary', 'sm')
    })
    
    it('should render with large size', () => {
      render(<Button size="lg">Large Button</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'primary', 'lg')
    })
  })
  
  describe('States', () => {
    it('should be disabled when disabled prop is true', () => {
      render(<Button disabled>Disabled Button</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toBeDisabled()
      expect(button).toHaveClass('button', 'primary', 'md')
    })
    
    it('should show loading state', () => {
      render(<Button loading>Save</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toBeDisabled()
      expect(button).toHaveAttribute('aria-busy', 'true')
      expect(button).toHaveClass('button', 'primary', 'md', 'loading')
      
      // Original text should be visible in span
      expect(screen.getByText('Save')).toBeInTheDocument()
    })
    
    it('should be disabled when loading', () => {
      render(<Button loading>Processing</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toBeDisabled()
      expect(button).toHaveClass('button', 'primary', 'md', 'loading')
    })
  })
  
  describe('Interactions', () => {
    it('should handle click events', async () => {
      const handleClick = vi.fn()
      const { user } = setup(<Button onClick={handleClick}>Click me</Button>)
      
      const button = screen.getByRole('button')
      await user.click(button)
      
      expect(handleClick).toHaveBeenCalledTimes(1)
    })
    
    it('should not trigger click when disabled', async () => {
      const handleClick = vi.fn()
      const { user } = setup(
        <Button disabled onClick={handleClick}>
          Disabled
        </Button>
      )
      
      const button = screen.getByRole('button')
      await user.click(button)
      
      expect(handleClick).not.toHaveBeenCalled()
    })
    
    it('should not trigger click when loading', async () => {
      const handleClick = vi.fn()
      const { user } = setup(
        <Button loading onClick={handleClick}>
          Loading
        </Button>
      )
      
      const button = screen.getByRole('button')
      await user.click(button)
      
      expect(handleClick).not.toHaveBeenCalled()
    })
    
    it('should support keyboard navigation', async () => {
      const handleClick = vi.fn()
      const { user } = setup(<Button onClick={handleClick}>Press Enter</Button>)
      
      const button = screen.getByRole('button')
      button.focus()
      
      await user.keyboard('{Enter}')
      expect(handleClick).toHaveBeenCalledTimes(1)
      
      await user.keyboard(' ') // Space key
      expect(handleClick).toHaveBeenCalledTimes(2)
    })
  })
  
  describe('Styling', () => {
    it('should apply custom className', () => {
      render(<Button className="custom-class">Custom</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('custom-class')
    })
    
    it('should maintain base classes with custom className', () => {
      render(<Button className="ml-4">Styled</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'primary', 'md', 'ml-4')
    })
    
    it('should have focus styles', () => {
      render(<Button>Focus me</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'primary', 'md')
      // Focus styles are handled by CSS, not classes
    })
  })
  
  describe('Accessibility', () => {
    it('should have proper ARIA attributes when loading', () => {
      render(<Button loading>Saving...</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveAttribute('aria-busy', 'true')
    })
    
    it('should not have aria-busy when not loading', () => {
      render(<Button>Normal Button</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveAttribute('aria-busy', 'false')
    })
    
    it('should support aria-label', () => {
      render(<Button aria-label="Close dialog">X</Button>)
      
      const button = screen.getByRole('button', { name: /close dialog/i })
      expect(button).toBeInTheDocument()
    })
    
    it('should have proper loading aria attributes', () => {
      render(<Button loading>Loading</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveAttribute('aria-busy', 'true')
      expect(button).toHaveAttribute('aria-label', '처리 중')
      // Loading spinner is created via CSS ::after pseudo-element
    })
  })
  
  describe('Props Forwarding', () => {
    it('should forward HTML button attributes', () => {
      render(
        <Button
          type="submit"
          form="test-form"
          name="submit-button"
          value="submit"
        >
          Submit
        </Button>
      )
      
      const button = screen.getByRole('button')
      expect(button).toHaveAttribute('type', 'submit')
      expect(button).toHaveAttribute('form', 'test-form')
      expect(button).toHaveAttribute('name', 'submit-button')
      expect(button).toHaveAttribute('value', 'submit')
    })
    
    it('should forward data attributes', () => {
      render(<Button data-testid="custom-button">Test</Button>)
      
      const button = screen.getByTestId('custom-button')
      expect(button).toBeInTheDocument()
    })
  })
  
  describe('Performance', () => {
    it('should not re-render unnecessarily', () => {
      const { rerender } = render(<Button>Static Button</Button>)
      
      const button = screen.getByRole('button')
      const initialHTML = button.outerHTML
      
      // Re-render with same props
      rerender(<Button>Static Button</Button>)
      
      expect(button.outerHTML).toBe(initialHTML)
    })
  })
  
  describe('Additional Variants', () => {
    it('should render with outline variant', () => {
      render(<Button variant="outline">Outline Button</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'outline', 'md')
    })
    
    it('should render with ghost variant', () => {
      render(<Button variant="ghost">Ghost Button</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'ghost', 'md')
    })
    
    it('should render with fullWidth prop', () => {
      render(<Button fullWidth>Full Width</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'primary', 'md', 'fullWidth')
    })
    
    it('should render with icon support', () => {
      render(<Button icon={<span>icon</span>}>With Icon</Button>)
      
      const button = screen.getByRole('button')
      expect(button).toHaveClass('button', 'primary', 'md', 'icon')
      expect(screen.getByText('icon')).toBeInTheDocument()
      expect(screen.getByText('With Icon')).toBeInTheDocument()
    })
  })
})