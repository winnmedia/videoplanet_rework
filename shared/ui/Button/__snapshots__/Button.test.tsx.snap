// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > Snapshots > should match snapshot for loading state 1`] = `
<button
  aria-busy="true"
  class="font-medium rounded-lg transition-colors focus:outline-none focus:ring-2 bg-blue-600 text-white hover:bg-blue-700 focus:ring-blue-500 px-4 py-2 text-base opacity-50 cursor-not-allowed "
  disabled=""
>
  <span
    class="inline-flex items-center"
  >
    <svg
      aria-hidden="true"
      class="animate-spin -ml-1 mr-2 h-4 w-4"
      fill="none"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        class="opacity-25"
        cx="12"
        cy="12"
        r="10"
        stroke="currentColor"
        stroke-width="4"
      />
      <path
        class="opacity-75"
        d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
        fill="currentColor"
      />
    </svg>
    Loading...
  </span>
</button>
`;

exports[`Button Component > Snapshots > should match snapshot for primary variant 1`] = `
<button
  aria-busy="false"
  class="font-medium rounded-lg transition-colors focus:outline-none focus:ring-2 bg-blue-600 text-white hover:bg-blue-700 focus:ring-blue-500 px-4 py-2 text-base  "
>
  Primary
</button>
`;
